### DISCUSSION TITLE
Coroutines Interop

### DISCUSSION THREAD

**Discussion by Intektor** [arkivanov/Decompose#46] [Q&A] [answered]
I was wondering why you use your own State implementation instead of using MutableStateFlow. Is there any benefit in using your State implementation, as it looks like it's just an observer pattern? As I really like to use flows, I was also asking myself if there is something similar to the Android viewModelScope, a componentScope maybe, that is cancelled when the component is killed.

**Comment by arkivanov** [comment, marked as answer]
Hello and thanks for the question.

> I was wondering why you use your own State implementation instead of using MutableStateFlow

This is required to avoid direct dependency on `kotlinx.coroutines`. One may prefer another library (e.g. [Reaktive](https://github.com/badoo/Reaktive) or RxJava). Or if Jetpack Compose is used as UI framework, then there is `State` and you may not need coroutines as well.

Another benefit of using `Value` is better interoperability with Swift. `Value` is a normal class with normal functions and callbacks - it can be easily exposed to Swift and subscribed/unsubscribed there.

Also please note, that `Value` is directly used only in `Router`. Usually its state is exposed to UI and subscribed without coroutines - e.g. `val routerState = root.routerState.subscribeAsState()`. For anything else you a free to use whatever you like.

> I was also asking myself if there is something similar to the Android viewModelScope, a componentScope maybe, that is cancelled when the component is killed

Since components are lifecycle-aware, you are free to create your own coroutine scope! 

```kotlin
fun LifecycleOwner.coroutineScope(context: CoroutineContext): CoroutineScope {
    val scope = CoroutineScope(context)
    lifecycle.doOnDestroy(scope::cancel)

    return scope
}
```

And then use it in a component as follows:

```kotlin
class SomeComponent(componentContext: ComponentContext): ComponentContext by componentContext {
    private val scope = coroutineScope(Dispatchers.Main)
}
```

**Reply by Burtan** [reply]
While you can get to the idea to do this yourself, you might add this to the documentation as it is mandatory when using coroutines.

**Reply by arkivanov** [reply]
Sure thing! Filed #150.

**Reply by arkivanov** [reply]
Updated the docs - https://arkivanov.github.io/Decompose/component/scopes/

**Reply by arkivanov** [reply]
Value should be also beneficial when exporting Decompose components to JavaScript (`binaries.library()`).
